phylokit/distance.py

```diff
@@ -0,0 +1,142 @@
+# Tree distance metrics.
+
+import numba
+import numpy as np
+
+
+@numba.njit(cache=True)
+def _branch_length(parent, time, u):
+    ret = 0
+    p = parent[u]
+    if p != -1:
+        ret = time[p] - time[u]
+    return ret
+
+
+def branch_length(tree, u):
+    """
+    Returns the length of the branch (in units of time) joining the specified
+    node to its parent.
+
+    :param int u: The node ID.
+    :return : The length of the branch.
+    :rtype : float
+    """
+    num_nodes = tree.tree_sequence.num_nodes
+    if u < 0 or u > num_nodes:
+        raise ValueError("Node index out of bounds")
+    return _branch_length(
+        tree.parent_array, tree.tree_sequence.tables.nodes.time, u
+    )
+
+
+@numba.njit(cache=True)
+def _mrca(parent, time, u, v):
+    tu = time[u]
+    tv = time[v]
+    while u != v:
+        if tu < tv:
+            u = parent[u]
+            if u == -1:
+                return -1
+            tu = time[u]
+        else:
+            v = parent[v]
+            if v == -1:
+                return -1
+            tv = time[v]
+    return u
+
+
+def mrca(tree, u, v):
+    """
+    Returns the most recent common ancestor of the specified nodes.
+
+    :param int `*args`: input node IDs, must be at least 2.
+    :return: The most recent common ancestor of input nodes.
+    :rtype: int
+    """
+    num_nodes = tree.tree_sequence.num_nodes
```

Comment:
Probably bettter for us to write a function like this:
```python
def check_node_bounds(tree, u):
     num_nodes = tree.parent_array.shape[0] - 1
     if u < 0 or u > num_nodes:
          raise ValueError(...)
```
we don't want to keep our dependencies on tskit to a minimum.

Reply:
I made the function a bit more general to check multiple nodes
```python
def check_node_bounds(tree, *args):
    """
    Checks that the specified node is within the tree.

    :param tskit.Tree tree: The tree to check.
    :param int `*args`: The node IDs to check.
    :raises ValueError: If any of the nodes are outside the tree.
    """
    num_nodes = tree.parent_array.shape[0] - 1
    for u in args:
        if u < 0 or u > num_nodes:
            raise ValueError("Node {} is not in the tree".format(u))
```

P.S. Should we make an inner function to return a boolean array to make it more general? (but it sounds pretty useless 😂)

Reply:
What you have is good :+1: